types/underscore/underscore-tests.ts

```diff
@@ -165,372 +165,423 @@ declare const level2UnionList: _.List<_.List<string | number>>;
 
 // creating a function that can determine if one object's property values matches another's
-var isUncleMoe = _.matches<Family>({ name: 'moe', relation: 'uncle' });
-var uncleMoe: Family = { name: 'moe', relation: 'uncle' };
-isUncleMoe(uncleMoe);
+{
+    const isUncleMoe = _.matches({ name: 'moe', relation: 'uncle' }); // $ExpectType Predicate<{ name: string; relation: string; }>
+    isUncleMoe({ name: 'moe', relation: 'uncle' }); // $ExpectType boolean
+}
 
 ///////////////////////////////////////////////////////////////////////////////////////
 
 // retrieving the first item in an array
-_.first([5, 4, 3, 2, 1]);
+_.first(numberArray); // $ExpectType number | undefined
 
 // retrieving all but the last element in an array
-_.initial([5, 4, 3, 2, 1]);
+_.initial(numberArray); // $ExpectType number[]
 
 // retrieving the last two elements in an array
-_.last([5, 4, 3, 2, 1], 2);
+_.last(numberArray, 2); // $ExpectType number[]
 
 // retrieving all but the first two elements in an array
-_.rest([5, 4, 3, 2, 1], 2);
+_.rest(numberArray, 2); // $ExpectType number[]
 
 // removing falsy values
-_.compact([0, 1, false, 2, '', 3, undefined]);
+_.compact([0, 1, false, 2, '', '3', undefined]); // $ExpectType (string | number | true)[]
 
 // deep flattening an array
-_.flatten([1, [2], [3, [[4]]]]);
+_.flatten([[[1, 2], [3]], [[4, 5]]]); // $ExpectType any[]
 
 // shallow flattening an array
-_.flatten([1, [2], [3, [[4]]]], true);
+_.flatten([[[1, 2], [3]], [[4, 5]]], true); // $ExpectType number[][]
 
 // excluding values from a list
-_.without([1, 2, 1, 0, 3, 1, 4], 0, 1);
+_.without([1, 2, 1, 0, 3, 1, 4], 0, 1); // $ExpectType number[]
 
 // computing the union of several sets
-_.union([1, 2, 3], [101, 2, 1, 10], [2, 1]);
+_.union([1, 2, 3], [101, 2, 1, 10], [2, 1]); // $ExpectType number[]
 
 // computing the intersection of several sets
-_.intersection([1, 2, 3], [101, 2, 1, 10], [2, 1]);
+_.intersection([1, 2, 3], [101, 2, 1, 10], [2, 1]); // $ExpectType number[]
 
 // computing the difference one set and other sets
-_.difference([1, 2, 3, 4, 5], [5, 2, 10]);
+_.difference([1, 2, 3, 4, 5], [5, 2, 10]); // $ExpectType number[]
 
 // determining the unique values in an array
-_.uniq([1, 2, 1, 3, 1, 4]);
+_.uniq([1, 2, 1, 3, 1, 4]); // $ExpectType number[]
 
 // merging together lists of values at a set of positions
-_.zip(['moe', 'larry', 'curly'], [30, 40, 50], [true, false, false]);
+_.zip(['moe', 'larry', 'curly'], [30, 40, 50], [true, false, false]); // $ExpectType any[][]
 
 // creating an object from a set of keys and a set of values
-var r = _.object(['moe', 'larry', 'curly'], [30, 40, 50]);
+_.object(['moe', 'larry', 'curly'], [30, 40, 50]); // $ExpectType Dictionary<number | undefined>
 
 // creating an object from a set of key-value pairs
-_.object([['moe', 30], ['larry', 40], ['curly', 50]]);
+_.object([['moe', 30], ['larry', 40], ['curly', 50]] as [string, number][]); // $ExpectType Dictionary<number>
```

Comment:
TS ends up interpreting the argument as `(string | number)[]` and since it then can't tell what position the types are in it ends up resulting in `Dictionary<string | number>` to be safe. Since the PR in which I added these I've observed functions like `Promise.all` provide an overload that expects a readonly tuple and given an array that is initialized as part of the call itself will infer the inline array's type as a tuple, so I could add that to remove the necessity of this assertion.

The below update to overload signatures causes the type assertion in this example to be unnecessary:
```TS
object<TList extends List<string | number>, TValue>(
    list: TList,
    values: List<TValue>
): Dictionary<TValue | undefined>;
object<TList extends List<Readonly<[string, any]>>>(
    list: TList
): Dictionary<PairValue<TypeOfList<TList>>>;
object<TList extends List<List<any>>>(
    list: TList
): Dictionary<PairValue<TypeOfList<TList>>>;
```

That being said, this only works for arrays that are being declared directly within the function call, and I feel like this is unlikely to be called that way in real code since this use case requires all property names and values to be known at the time of the call and in cases like that one might as well create an object directly without bothering to do so through this function. It might be best to leave this as-is to make it clearer that having an array typed as a tuple will provide better results here and also to avoid adding the cost of an overload that covers a very unlikely use case. To make it clearer that the example call here is not the best real-world usage, I could also replace the example declared inline here with a reference to a `declare const nameAgePairs: [string, number][];`.

Reply:
I'm happy enough with this explanation, but if you prefer forward-declaring `nameAgePairs`, that's also fine by me.